File: contact-management/php/contact_management.php
<?php
/**
 * SendSeven API - Contact Management Example (PHP)
 *
 * Demonstrates CRUD operations for contacts using the SendSeven API.
 */

// Load environment variables from .env file (simple implementation)
function loadEnv(string $path): void {
    if (!file_exists($path)) {
        return;
    }
    $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        if (strpos(trim($line), '#') === 0) {
            continue;
        }
        if (strpos($line, '=') !== false) {
            list($name, $value) = explode('=', $line, 2);
            putenv(trim($name) . '=' . trim($value));
        }
    }
}

loadEnv(__DIR__ . '/.env');

// Configuration from environment
$API_TOKEN = getenv('SENDSEVEN_API_TOKEN');
$TENANT_ID = getenv('SENDSEVEN_TENANT_ID');
$API_URL = getenv('SENDSEVEN_API_URL') ?: 'https://api.sendseven.com/api/v1';

/**
 * Get common headers for API requests.
 */
function getHeaders(): array {
    global $API_TOKEN, $TENANT_ID;
    return [
        'Authorization: Bearer ' . $API_TOKEN,
        'X-Tenant-ID: ' . $TENANT_ID,
        'Content-Type: application/json',
    ];
}

/**
 * Make an HTTP request to the API.
 */
function makeRequest(string $method, string $endpoint, ?array $data = null): array {
    global $API_URL;

    $url = $API_URL . $endpoint;
    $ch = curl_init($url);

    $options = [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ];

    switch (strtoupper($method)) {
        case 'POST':
            $options[CURLOPT_POST] = true;
            if ($data !== null) {
                $options[CURLOPT_POSTFIELDS] = json_encode($data);
            }
            break;
        case 'PUT':
            $options[CURLOPT_CUSTOMREQUEST] = 'PUT';
            if ($data !== null) {
                $options[CURLOPT_POSTFIELDS] = json_encode($data);
            }
            break;
        case 'DELETE':
            $options[CURLOPT_CUSTOMREQUEST] = 'DELETE';
            break;
        case 'GET':
        default:
            break;
    }

    curl_setopt_array($ch, $options);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Create a new contact.
 */
function createContact(array $contactData): array {
    return makeRequest('POST', '/contacts', $contactData);
}

/**
 * List contacts with pagination.
 */
function listContacts(int $page = 1, int $pageSize = 20): array {
    return makeRequest('GET', "/contacts?page=$page&page_size=$pageSize");
}

/**
 * Get a single contact by ID.
 */
function getContact(string $contactId): array {
    return makeRequest('GET', "/contacts/$contactId");
}

/**
 * Update an existing contact.
 */
function updateContact(string $contactId, array $contactData): array {
    return makeRequest('PUT', "/contacts/$contactId", $contactData);
}

/**
 * Delete a contact.
 */
function deleteContact(string $contactId): array {
    return makeRequest('DELETE', "/contacts/$contactId");
}

// Main execution
function main(): void {
    global $API_TOKEN, $TENANT_ID;

    // Validate configuration
    if (!$API_TOKEN) {
        echo "Error: SENDSEVEN_API_TOKEN environment variable is required\n";
        exit(1);
    }

    if (!$TENANT_ID) {
        echo "Error: SENDSEVEN_TENANT_ID environment variable is required\n";
        exit(1);
    }

    echo "SendSeven Contact Management Example\n";
    echo str_repeat('=', 40) . "\n";

    try {
        // 1. Create a new contact
        echo "\n1. Creating a new contact...\n";
        $contact = createContact([
            'phone_number' => '+1234567890',
            'email' => 'john.doe@example.com',
            'first_name' => 'John',
            'last_name' => 'Doe',
            'company' => 'Acme Inc',
        ]);
        $contactId = $contact['id'];
        echo "   Created contact: {$contactId}\n";
        echo "   Name: {$contact['first_name']} {$contact['last_name']}\n";
        echo "   Email: {$contact['email']}\n";
        echo "   Phone: {$contact['phone_number']}\n";

        // 2. List contacts
        echo "\n2. Listing contacts...\n";
        $contactsResponse = listContacts(1, 10);
        echo "   Total contacts: {$contactsResponse['pagination']['total']}\n";
        echo "   Page {$contactsResponse['pagination']['page']} of {$contactsResponse['pagination']['total_pages']}\n";
        foreach (array_slice($contactsResponse['items'], 0, 3) as $c) {
            $name = trim(($c['first_name'] ?? '') . ' ' . ($c['last_name'] ?? '')) ?: 'Unnamed';
            echo "   - {$c['id']}: $name\n";
        }

        // 3. Get single contact
        echo "\n3. Getting contact {$contactId}...\n";
        $fetchedContact = getContact($contactId);
        echo "   ID: {$fetchedContact['id']}\n";
        echo "   Name: {$fetchedContact['first_name']} {$fetchedContact['last_name']}\n";
        echo "   Company: {$fetchedContact['company']}\n";

        // 4. Update contact
        echo "\n4. Updating contact {$contactId}...\n";
        $updatedContact = updateContact($contactId, [
            'first_name' => 'Jane',
            'company' => 'New Company Inc',
        ]);
        echo "   Updated name: {$updatedContact['first_name']} {$updatedContact['last_name']}\n";
        echo "   Updated company: {$updatedContact['company']}\n";

        // 5. Delete contact
        echo "\n5. Deleting contact {$contactId}...\n";
        $deleteResult = deleteContact($contactId);
        $deleted = isset($deleteResult['success']) ? ($deleteResult['success'] ? 'true' : 'false') : 'true';
        echo "   Deleted: $deleted\n";

        echo "\n" . str_repeat('=', 40) . "\n";
        echo "All operations completed successfully!\n";
    } catch (Exception $e) {
        echo "\nError: " . $e->getMessage() . "\n";
        exit(1);
    }
}

main();

File: contact-management/python/contact_management.py
#!/usr/bin/env python3
"""
SendSeven API - Contact Management Example

Demonstrates CRUD operations for contacts using the SendSeven API.
"""

import os
import requests
from dotenv import load_dotenv
from typing import Optional

# Load environment variables from .env file
load_dotenv()

# Configuration from environment
API_TOKEN = os.getenv("SENDSEVEN_API_TOKEN")
TENANT_ID = os.getenv("SENDSEVEN_TENANT_ID")
API_URL = os.getenv("SENDSEVEN_API_URL", "https://api.sendseven.com/api/v1")


def get_headers() -> dict:
    """Get common headers for API requests."""
    return {
        "Authorization": f"Bearer {API_TOKEN}",  # Bearer token authentication
        "X-Tenant-ID": TENANT_ID,
        "Content-Type": "application/json",
    }


def create_contact(
    phone: Optional[str] = None,
    email: Optional[str] = None,
    name: Optional[str] = None,
) -> dict:
    """
    Create a new contact.

    Args:
        phone: Phone number in E.164 format
        email: Email address
        name: Contact's full name

    Returns:
        dict: The created contact object

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/contacts"

    payload = {}
    if phone:
        payload["phone"] = phone
    if email:
        payload["email"] = email
    if name:
        payload["name"] = name

    response = requests.post(url, json=payload, headers=get_headers())
    response.raise_for_status()

    return response.json()


def list_contacts(page: int = 1, page_size: int = 20) -> dict:
    """
    List contacts with pagination.

    Args:
        page: Page number (1-indexed)
        page_size: Number of contacts per page

    Returns:
        dict: Paginated response with items and pagination info

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/contacts"
    params = {"page": page, "page_size": page_size}

    response = requests.get(url, params=params, headers=get_headers())
    response.raise_for_status()

    return response.json()


def get_contact(contact_id: str) -> dict:
    """
    Get a single contact by ID.

    Args:
        contact_id: The contact's UUID

    Returns:
        dict: The contact object

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/contacts/{contact_id}"

    response = requests.get(url, headers=get_headers())
    response.raise_for_status()

    return response.json()


def update_contact(
    contact_id: str,
    phone: Optional[str] = None,
    email: Optional[str] = None,
    name: Optional[str] = None,
) -> dict:
    """
    Update an existing contact.

    Args:
        contact_id: The contact's UUID
        phone: Phone number in E.164 format
        email: Email address
        name: Contact's full name

    Returns:
        dict: The updated contact object

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/contacts/{contact_id}"

    payload = {}
    if phone is not None:
        payload["phone"] = phone
    if email is not None:
        payload["email"] = email
    if name is not None:
        payload["name"] = name

    response = requests.put(url, json=payload, headers=get_headers())
    response.raise_for_status()

    return response.json()


def delete_contact(contact_id: str) -> dict:
    """
    Delete a contact.

    Args:
        contact_id: The contact's UUID

    Returns:
        dict: Deletion confirmation

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/contacts/{contact_id}"

    response = requests.delete(url, headers=get_headers())
    response.raise_for_status()

    return response.json()


def main():
    # Validate configuration
    if not API_TOKEN:
        print("Error: SENDSEVEN_API_TOKEN environment variable is required")
        return

    if not TENANT_ID:
        print("Error: SENDSEVEN_TENANT_ID environment variable is required")
        return

    print("SendSeven Contact Management Example")
    print("=" * 40)

    try:
        # 1. Create a new contact
        print("\n1. Creating a new contact...")
        contact = create_contact(
            phone="+1234567890",
            email="john.doe@example.com",
            name="John Doe"
        )
        contact_id = contact["id"]
        print(f"   Created contact: {contact_id}")
        print(f"   Name: {contact.get('name')}")
        print(f"   Email: {contact.get('email')}")
        print(f"   Phone: {contact.get('phone')}")

        # 2. List contacts
        print("\n2. Listing contacts...")
        contacts_response = list_contacts(page=1, page_size=10)
        print(f"   Total contacts: {contacts_response['pagination']['total']}")
        print(f"   Page {contacts_response['pagination']['page']} of {contacts_response['pagination']['total_pages']}")
        for c in contacts_response["items"][:3]:  # Show first 3
            # Use name or fall back to identifiers
            name = c.get('name') or c.get('phone') or c.get('email') or f"Contact {c['id'][:8]}..."
            print(f"   - {c['id']}: {name}")

        # 3. Get single contact
        print(f"\n3. Getting contact {contact_id}...")
        fetched_contact = get_contact(contact_id)
        print(f"   ID: {fetched_contact['id']}")
        print(f"   Name: {fetched_contact.get('name')}")

        # 4. Update contact
        print(f"\n4. Updating contact {contact_id}...")
        updated_contact = update_contact(
            contact_id,
            name="Jane Doe"
        )
        print(f"   Updated name: {updated_contact.get('name')}")

        # 5. Delete contact
        print(f"\n5. Deleting contact {contact_id}...")
        delete_result = delete_contact(contact_id)
        print(f"   Deleted: {delete_result.get('success', True)}")

        print("\n" + "=" * 40)
        print("All operations completed successfully!")

    except requests.HTTPError as e:
        print(f"\nAPI Error: {e.response.status_code}")
        print(f"Response: {e.response.text}")
    except Exception as e:
        print(f"\nError: {e}")


if __name__ == "__main__":
    main()
